### ISSUE TITLE
`COPY` from s3: invalid data (Must reach the buffer end) (while in processor thread 0)

### ISSUE THREAD

**Issue by BohuTANG** [datafuselabs/databend#6061]
**Summary**

## Deploy
Deploy databend on aws ec2: https://databend.rs/doc/deploy/s3
version: v0.7.91-nightly

## Ontime table
https://databend.rs/doc/learn/analyze-ontime-with-databend-on-ec2-and-s3#22-create-ontime-table

## Copy csv data from aws s3 public bucket
```
mysql> COPY INTO ontime   FROM 's3://databendcloud/ontime/' pattern = 'ontime_.*csv'   FILE_FORMAT = (type = "CSV" field_delimiter = ','  record_delimiter = '\n' skip_header = 1) size_limit=10;
ERROR 1105 (HY000): Code: 4000, displayText = invalid data (Must reach the buffer end) (while in processor thread 0).
```

server log:
```
ble_virtual_host_style: true }
2022-06-19T13:44:09.164384Z  INFO opendal::services::s3::backend: backend use root /ontime/
2022-06-19T13:44:09.164672Z  INFO opendal::services::s3::backend: backend build started: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.164711Z  INFO opendal::services::s3::backend: backend use root /ontime/
2022-06-19T13:44:09.166519Z  INFO opendal::services::s3::backend: backend build started: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.166567Z  INFO opendal::services::s3::backend: backend use root /ontime/
2022-06-19T13:44:09.167255Z  INFO opendal::services::s3::backend: backend build started: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.167296Z  INFO opendal::services::s3::backend: backend use root /ontime/
2022-06-19T13:44:09.168357Z  INFO opendal::services::s3::backend: backend build started: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.168398Z  INFO opendal::services::s3::backend: backend use root /ontime/
2022-06-19T13:44:09.168535Z  INFO opendal::services::s3::backend: backend build started: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.168576Z  INFO opendal::services::s3::backend: backend use root /ontime/
2022-06-19T13:44:09.169645Z  INFO opendal::services::s3::backend: backend build started: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.169687Z  INFO opendal::services::s3::backend: backend use root /ontime/
2022-06-19T13:44:09.170153Z  INFO opendal::services::s3::backend: backend build started: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.170193Z  INFO opendal::services::s3::backend: backend use root /ontime/
2022-06-19T13:44:09.170865Z  INFO opendal::services::s3::backend: backend build started: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.170904Z  INFO opendal::services::s3::backend: backend use root /ontime/
2022-06-19T13:44:09.170940Z  INFO opendal::services::s3::backend: backend build started: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.170981Z  INFO opendal::services::s3::backend: backend use root /ontime/
2022-06-19T13:44:09.178800Z  INFO opendal::services::s3::backend: backend build started: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.178843Z  INFO opendal::services::s3::backend: backend use root /ontime/
2022-06-19T13:44:09.749245Z  INFO opendal::services::s3::backend: backend build finished: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.749301Z  INFO databend_query::storages::stage::stage_source: Read stage s3://databendcloud/ontime/ file ontime_000.csv via compression None
2022-06-19T13:44:09.751558Z  INFO opendal::services::s3::backend: backend build finished: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.751596Z  INFO databend_query::storages::stage::stage_source: Read stage s3://databendcloud/ontime/ file ontime_005.csv via compression None
2022-06-19T13:44:09.760501Z  INFO opendal::services::s3::backend: backend build finished: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.760547Z  INFO databend_query::storages::stage::stage_source: Read stage s3://databendcloud/ontime/ file ontime_003.csv via compression None
2022-06-19T13:44:09.761451Z  INFO opendal::services::s3::backend: backend build finished: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.761491Z  INFO databend_query::storages::stage::stage_source: Read stage s3://databendcloud/ontime/ file ontime_009.csv via compression None
2022-06-19T13:44:09.761628Z  INFO opendal::services::s3::backend: backend build finished: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.761658Z  INFO databend_query::storages::stage::stage_source: Read stage s3://databendcloud/ontime/ file ontime_006.csv via compression None
2022-06-19T13:44:09.761804Z  INFO opendal::services::s3::backend: backend build finished: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.761835Z  INFO databend_query::storages::stage::stage_source: Read stage s3://databendcloud/ontime/ file ontime_014.csv via compression None
2022-06-19T13:44:09.804909Z  INFO databend_query::storages::stage::stage_source: Read stage s3://databendcloud/ontime/ file ontime_028.csv via compression None
2022-06-19T13:44:09.805834Z  INFO opendal::services::s3::backend: backend build finished: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.805867Z  INFO databend_query::storages::stage::stage_source: Read stage s3://databendcloud/ontime/ file ontime_017.csv via compression None
2022-06-19T13:44:09.807073Z  INFO opendal::services::s3::backend: backend build finished: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.807114Z  INFO databend_query::storages::stage::stage_source: Read stage s3://databendcloud/ontime/ file ontime_027.csv via compression None
2022-06-19T13:44:09.807601Z  INFO opendal::services::s3::backend: backend build finished: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.807635Z  INFO databend_query::storages::stage::stage_source: Read stage s3://databendcloud/ontime/ file ontime_020.csv via compression None
2022-06-19T13:44:09.808245Z  INFO opendal::services::s3::backend: backend build finished: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.808274Z  INFO databend_query::storages::stage::stage_source: Read stage s3://databendcloud/ontime/ file ontime_025.csv via compression None
2022-06-19T13:44:09.808791Z  INFO opendal::services::s3::backend: backend build finished: Builder { root: Some("/ontime/"), bucket: "databendcloud", endpoint: Some("https://s3.amazonaws.com"), region: None, disable_credential_loader: true, enable_virtual_host_style: true }
2022-06-19T13:44:09.808836Z  INFO databend_query::storages::stage::stage_source: Read stage s3://databendcloud/ontime/ file ontime_031.csv via compression None
2022-06-19T13:44:09.871014Z ERROR databend_query::servers::mysql::writers::query_result_writer: OnQuery Error: Code: 4000, displayText = invalid data (Must reach the buffer end) (while in processor thread 0).
```

**Comment by BohuTANG** [comment]
The error `Must reach the buffer end` is a bad log for humans; we don't know what's error is in the underlying. cc @sundy-li

**Comment by sundy-li** [comment]
It means we did not allow trailing chars in deserialization.
```
fn must_eof(&mut self) -> Result<()> {
        let buffer = self.fill_buf()?;
        if !buffer.is_empty() {
            return Err(std::io::Error::new(
                ErrorKind::InvalidData,
                "Must reach the buffer end",
            ));
        }
        Ok(())
}
```

**Comment by BohuTANG** [comment]
The `must_eof` is called by some types
![image](https://user-images.githubusercontent.com/172204/174506520-ec0b5692-32b3-48c8-af6f-f4a3c06f5fbe.png)

If we got the `Must reach the buffer end`, we next want to know which type caused the error: date?number?timestamp?
If we can print the file line where the error is, it would be great, like:
```
"Invalid Data: Must reach the buffer end, [parse number data type], file:xx.csv,  line: xx "
```

Then we can easy to check the line of the file.

**Comment by BohuTANG** [comment]
This issue is caused by `field_delimiter` is not correct(is `\t` not `,`), this is OK:
```
COPY INTO ontime   FROM 's3://databendcloud/ontime/' pattern = 'ontime_.*csv'   FILE_FORMAT = (type = "CSV" field_delimiter = '\t'  record_delimiter = '\n' skip_header = 1) size_limit=10;
```
Can we improve the error message to be more friendly? I believe that most people who see this error do not know what went wrong.

**Comment by sundy-li** [comment]
Fixed in #6530
